### PR TITLE
Raven db 7139

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

File: src/Raven.Client/Properties/VersionInfo.cs [hunk 1]
```diff
@@ -3,7 +3,7 @@
 using System.Reflection;
 using Raven.Client.Properties;
 
-[assembly: RavenVersion(Build = "40", CommitHash = "2850730", Version = "4.0", FullVersion = "4.0.0-custom-40")]
+[assembly: RavenVersion(Build = "40", CommitHash = "005256c", Version = "4.0", FullVersion = "4.0.0-custom-40")]
```
Comment (ayende) [inline src/Raven.Client/Properties/VersionInfo.cs:6]: Try to avoid committing these files

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 2]
```diff
@@ -156,17 +157,14 @@ public void ClusterOnDatabaseChanged(object sender, (string dbName, long index)
...
-            var cmd = new DynamicJsonValue
+            var cmd = new RemoveNodeFromDatabaseCommand(dbName)
             {
-                ["Type"] = nameof(RemoveNodeFromDatabaseCommand),
-                [nameof(RemoveNodeFromDatabaseCommand.DatabaseName)] = dbName,
-                [nameof(RemoveNodeFromDatabaseCommand.NodeTag)] = _serverStore.NodeTag
+                NodeTag = _serverStore.NodeTag
             };
             JsonOperationContext myContext;
             using (_serverStore.ContextPool.AllocateOperationContext(out myContext))
```
Comment (ayende) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: Why do we need the context here?

File: src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            var deletionInProgressStatus = HardDelete? DeletionInProgressStatus.HardDelete
+                : DeletionInProgressStatus.SoftDelete;
+            if (record.DeletionInProgress == null)
+            {
+                record.DeletionInProgress = new Dictionary<string, DeletionInProgressStatus>();
+            }
+            if (string.IsNullOrEmpty(FromNode) == false)
+            {
+                //TODO: maybe expose a way to issue errors when applying commands so we can atleast raise alerts
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs]: There is, throw an exception, it is handled.

File: src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                }
+                record.Topology.RemoveFromTopology(FromNode);
+
+                record.DeletionInProgress[FromNode] = deletionInProgressStatus;
+            }
+            else
+            {
+                var allNodes = record.Topology.Members.Select(m => m.NodeTag)
+                    .Concat(record.Topology.Promotables.Select(p => p.NodeTag));
+                    // TODO: we need to delete databases from watchers too but watcher nodeTag seems to be null
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Commands/DeleteDatabaseCommand.cs:46]: No, watchers are external to the topology

File: src/Raven.Client/Server/DatabaseTopology.cs [hunk 5]
```diff
@@ -241,7 +245,8 @@ public DynamicJsonValue ToJson()
         public void RemoveFromTopology(string delDbFromNode)
         {
             Members.RemoveAll(m => m.NodeTag == delDbFromNode);
-            Promotables.RemoveAll(p=> p.NodeTag == delDbFromNode);
+            Promotables.RemoveAll(p => p.NodeTag == delDbFromNode);
+            Watchers.RemoveAll(w => w.NodeTag == delDbFromNode);
```
Comment (ayende) [inline src/Raven.Client/Server/DatabaseTopology.cs]: Removing watcher is an explicit decision

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by talweiss1982** [comment]
fixed comments

**Pull request merged**